scrapy/settings/__init__.py

```diff
@@ -34,9 +40,20 @@ def __init__(self, value, priority):
 
     def set(self, value, priority):
         """Sets value if priority is higher or equal than current priority."""
-        if priority >= self.priority:
-            self.value = value
-            self.priority = priority
+        if isinstance(self.value, BaseSettings) and isinstance(value, Mapping):
+            # Ignore self.priority if self.value has per-key priorities
+            # XXX: Second condition allows completely replacing a BaseSettings
+            #      instance with a value of incompatible type (i.e. something
+            #      that does not map keys to values, e.g.a list). I included
+            #      this b/c the test_cmdline/__init__.py test uses a deprecated
+            #      configuration API (EXTENSIONS is a list) and fails otherwise.
+            #      Should it stay in?
```

Comment:
On second thought, now that only the default dicts are promoted, and given that these should never be replaced by any incompatible type, I will remove the `and isinstance(value, Mapping`). It also causes problems with passing json-encoded strings.